docs(api): tutorial.rst revisions for Flex (#13046)
OpenTrons/opentrons-api: api/docs/v2/tutorial.rst

To run your code, make sure that you've installed `Python 3 <https://wiki.python.org/moin/BeginnersGuide/Download>`_ and the `pip package installer <https://pip.pypa.io/en/stable/getting-started/>`_. You should write your code in your favorite plaintext editor or IDE and save it in a file with a ``.py`` extension, like ``dilution-tutorial.py``.

Hardware and Labware
^^^^^^^^^^^^^^^^^^^^

Before running a protocol, you’ll want to have the right kind of hardware and labware ready for your Flex or OT-2.

.. Placeholders for Flex content that needs links to online manual.

> - **Flex users** should review Chapter 2: Installation and Relocation (**LINK TBD**) in the instruction manual. Specifically, see the pipette information in the Instrument Installation and Calibration section (**LINK TBD**). You can use either a 1-channel or 8-channel pipette for this tutorial. Most Flex code examples will use a `1-channel, 1000 uL pipette <https://opentrons.com/products/flex-pipettes/>`_.

Reviewer: Let's just link `instruction manual` for now. And to mirror what we have for OT-2 (and to be as specific as possible), we can link to the shop page for the 1-channel. I've also updated it to use the official short name. 
```suggestion
- **Flex users** should review Chapter 2: Installation and Relocation in the `instruction manual <https://insights.opentrons.com/hubfs/Products/Flex/Opentrons%20Flex%20Manual.pdf>`_. Specifically, see the pipette information in the Instrument Installation and Calibration section. You can use either a 1-channel or 8-channel pipette for this tutorial. Most Flex code examples will use a `Flex 1-Channel 1000 μL pipette <https://shop.opentrons.com/opentrons-flex-1-channel-pipette/>`_.
```

- **OT-2 users** should review the robot setup and pipette information on the `Get Started page <https://support.opentrons.com/s/ot2-get-started>`_. Specifically, see `attaching pipettes <https://support.opentrons.com/s/article/Get-started-Attach-pipettes>`_ and `initial calibration <https://support.opentrons.com/s/article/Get-started-Calibrate-the-deck>`_. You can use either a single-channel or 8-channel pipette for this tutorial. Most OT-2 code examples will use a `P300 Single-Channel GEN2 <https://shop.opentrons.com/single-channel-electronic-pipette-p20/>`_ pipette.

> The Flex and OT-2 use the same labware for serial dilution. The tutorial code will use the following labware definitions, but as long as you have labware of each type you can modify the code to run with your labware.

Reviewer: the tip racks are different for Flex/OT-2. we should either factor that out to the robot-specific sections above or load all that info into the one bullet below.

Author: Ohhhhh, right! The 200 ul Flex tips are in the shop site but not on the labware page. URLs for the Flex and OT-2 will be different: Flex goes to "shop.opentrons.com" and OT-2 will go to "labware.opentrons.com". 

We'll use the correct API load name for each: 

- opentrons_flex_96_tiprack_200ul
- opentrons_96_tiprack_300ul

    from opentrons import protocol_api

> .. Suggest simplifying this paragraph

Reviewer: looks good, can delete the temporary comment

Author: Will remove the comment.

> Throughout this documentation, you’ll see protocols that begin with the ``opentrons import`` statement shown above. It identifies your code as an Opentrons protocol. This statement is not required, but including it is a good practice and allows most code editors to provide helpful autocomplete suggestions. 

Reviewer: `opentrons import` is not idiomatic for python users. either
```the ``import`` statement```
or
```the ``from opentrons import`` statement```

Author: Will replace with "the `import` statement" as in the original.

        'author': 'New API User'
        }

.. note::
    If you have a Flex, or an OT-2 with API v2.15 (or higher), you set the API version in the ``requirements`` code block, not the ``metadata`` code block. See the Requirements section below.

.. With your metadata defined, you can move on to creating the ``run()`` function for your protocol.

Requirements
^^^^^^^^^^^^
.. My understanding of the requirements code block. Needs verification.

> The ``requirements`` code block appears after the ``metadata`` code block in a Python protocol. It defines the robot model (e.g. ``Flex`` or ``OT-2``) and the API version you're using.

Reviewer: technically these blocks can go just about anywhere. but this is good form, so saying "it goes here" is 👍

Author: Max suggested putting the `requirements` first, before `metadata` but also indicated order doesn't matter.

Flex users must specify the API version in the ``requirements`` instead of in the ``metadata``. The minimum required API version for Flex is v2.15. For example:

.. code-block:: python
    :substitutions:

    # Flex
    requirements = {"robotType": "Flex", "apiLevel": "2.15"}

> OT-2 users working with API v2.15 (or higher) also need to specify the API version in the ``requirements`` instead of in the ``metadata``. For example:

Reviewer: as of right now, you can specify any `apiLevel` in metadata. the only required key in requirements is `robotType`. but we should check if that's the long-term plan.

        
        Now the robot will expect to find labware in a configuration that looks like this:

        **PLACEHOLDER FOR FLEX DECK IMAGE**
    
    .. tab:: OT-2

       Here’s how to load the labware into OT-2 slots 1, 2, and 3 (repeating the ``def`` statement from above to show proper indenting):
        
       .. code-block:: python
           :substitutions:

           def run(protocol: protocol_api.ProtocolContext):
               tips = protocol.load_labware('opentrons_96_tiprack_300ul', 1)
               reservoir = protocol.load_labware('nest_12_reservoir_15ml', 2)
               plate = protocol.load_labware('nest_96_wellplate_200ul_flat', 3)
       
       If you’re using a different model of labware, find its name in the Labware Library and replace it in your code.
       
       Now the robot will expect to find labware in a configuration that looks like this:

       .. image:: ../img/tutorial/initial-deck-map.png
           :scale: 50 %
           :name: Initial Deck State
           :align: center
           :alt: OT-2 deck map with a tip rack in slot 1, reservoir in slot 2, and well plate in slot 3.

> You may notice that these deck maps don't show where the liquids will be at the start of the protocol. Liquid definitions aren’t required in Python protocols, unlike protocols made in `Protocol Designer <https://designer.opentrons.com/>`_. (Sneak peek: you’ll put the diluent in column 1 of the reservoir and the solution in column 2 of the reservoir.)

Reviewer: i think it's too advanced to put liquid defs in the tutorial, but we should link to that section of the Python docs. (it didn't exist when the Tutorial was first published!)

Author: Good point! Will revise and update with [Labelling Liquids in Wells](https://docs.opentrons.com/v2/new_labware.html#labeling-liquids-in-wells).

Pipettes
--------

> Next you’ll specify what pipette to use in the protocol. Loading a pipette is done with the :py:meth:`.load_instrument` method, which takes three arguments: the name of the pipette, the mount it’s installed in, and the tip racks it should use when performing transfers. Load whatever pipette you have installed in your robot by using its :ref:`standard pipette name <new-pipette-models>`. Here’s how to load Flex and OT-2 pipettes installed in the left mount:

Reviewer: ```suggestion
Next you’ll specify what pipette to use in the protocol. Loading a pipette is done with the :py:meth:`.load_instrument` method, which takes three arguments: the name of the pipette, the mount it’s installed in, and the tip racks it should use when performing transfers. Load whatever pipette you have installed in your robot by using its :ref:`standard pipette name <new-pipette-models>`. Here’s how to load the pipette in the left mount:
```
can simplify

.. code-block:: python

        # Flex
        p1000 = protocol.load_instrument('flex_1channel_1000', 'left', tip_racks[tips])

.. code-block:: python

        # OT-2
        p300 = protocol.load_instrument('p300_single_gen2', 'left', tip_racks=[tips])

> Since the pipette is so fundamental to the protocol, it might seem like you should have specified it first. But there’s a good reason why pipettes are loaded after labware: you need to have already loaded ``tips`` in order to tell the pipette to use it. And now you won’t have to reference ``tips`` again in your code — it’s assigned to the ``p1000`` (Flex) or ``p300`` (OT-2) pipette and the robot will know to use it when commanded to pick up tips.

Reviewer: if you think explaining the different variable names like this is awkward, we could use something like `left_pipette` instead for both robots.

Reviewer: this will have the added benefit of letting some of the code snippets below be robot-agnostic, instead of having to provide two one-line snippets that don't really differ in function at all.

Author: Ah, now I understand. Yes, instead of `p300` or `p1000`, let's use `left_pipette` for the variable name. And this transfers over to the animated gif. Will update.

Author: Updated with `left_pipette`

.. code-block:: python

        #Flex
        p1000.transfer(100, reservoir['A1'], plate.wells())

.. code-block:: python

        #OT-2
        p300.transfer(100, reservoir['A1'], plate.wells())

Breaking down these single lines of code shows the power of :ref:`complex commands <v2-complex-commands>`. The first argument is the amount to transfer to each destination, 100 µL. The second argument is the source, column 1 of the reservoir (which is still specified with grid-style coordinates as ``A1`` — a reservoir only has an A row). The third argument is the destination. Here, calling the :py:meth:`.wells` method of ``plate`` returns a list of *every well*, and the command will apply to all of them.

> .. Possible short text to explain the animation uses OT-2 code but applies to Flex as well?

Reviewer: ah, dang. i have the source file and could fairly easily update the animations.

Author: It isn't too late. Do you want to update the animation? 

Maybe that isn't required. Even though the code here uses the OT-2 pipette variable name, the concept is the same for both robots.  Creating 2 separate image files may be more work than needed to convey the idea showing stepwise transfer of liquids. Instead, we could try some disclaimer text like this:

"The following animated images use OT-2 code, but the concepts apply to how a Flex and OT-2 transfer liquids in this tutorial."

Something like that maybe? 

There are 5 animated gifs here. Making separate images for Flex creates a total of 10 images and also multiple tabs (if we put images in those). That's a lot of extra files to maintain and more tabs may make the page look busy.

Reviewer: i meant that if we went with a single pipette variable name, e.g. `left_pipette`, i could make one set of updated images. same overall complexity. just updating that little bit of text and re-exporting.

Author: Fixing this now. Change `p300` and `p1000` to `left_pipette` so we can use one code example for both robots and the animated image.

In Simulation
^^^^^^^^^^^^^

> Simulation doesn’t require having a robot to your computer. You just need to install the ``opentrons`` Python module using ``pip``. This will give you access to the ``opentrons_simulate`` command-line utility (``opentrons_simulate.exe`` on Windows). To see a text preview of the steps the OT-2 will take, ``cd`` to the directory where you saved your protocol file and run:

Reviewer: ```suggestion
Simulation doesn’t require having a robot connected to your computer. You just need to install the ``opentrons`` Python module using ``pip``. This will give you access to the ``opentrons_simulate`` command-line utility (``opentrons_simulate.exe`` on Windows). To see a text preview of the steps the OT-2 will take, ``cd`` to the directory where you saved your protocol file and run:
```

On a Robot
^^^^^^^^^^

> The simplest way to run your protocol on a Flex or OT-2 is to use the `Opentrons App <https://opentrons.com/ot-app>`_. Once you’ve installed the app and connected to your robot, navigate to the **Protocol** tab. Click **Choose File…** and open your protocol from the file picker. You should see “Protocol - Serial Dilution Tutorial” (or whatever ``protocolName`` you entered in the metadata) in a banner at the top of the page. 

Reviewer: on line 369 below: `Protocol Steps` are now called `Run Preview`. we should change the term here.

Author: Ok. Will change to `Run Preview`.
